### PR TITLE
[Reviewer: Ellie] Add timer heap implementation

### PR DESCRIPTION
**Pull request by rkd-msw**
This adds a timer heap implementation, for use as Chronos' extra heap and anywhere else we might need to queue up work for the future (e.g. it might be useful in your alarms work).

Fully UTed in cpp-common-test, PR there to follow.

### CODE REVIEW

**Comment by rkd-msw** [comment]
Hold off on reviewing this for now - @AlexHockey and @bossmc have suggested it would be simpler as a wrapper around http://www.boost.org/doc/libs/1_59_0/doc/html/boost/heap/binomial_heap.html.

**Comment by rkd-msw** [comment]
OK, ready for review.

File: include/timer_heap.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+  /// Heap which this timer is in, or NULL if it isn't currently in a heap.
+  ///
+  /// TimerHeap::insert is responsible for updating this field.
+  TimerHeap* _heap = nullptr;
+
+  // The current position of this timer in the heap's underlying array. The
+  // TimerHeap is responsible for keeping this up-to-date as it moves the
+  // timer around.
+  boost::heap::d_ary_heap<HeapableTimer*,
```
Comment (eleanor-merry) [inline include/timer_heap.h]: Not convinced by this alignment

**Comment by eleanor-merry** [comment]
LGTM

**Pull request merged**